syncstorage-postgres/src/db/db_impl.rs

```diff
@@ -50,18 +50,105 @@ impl Db for PgDb {
         Ok(true)
     }
 
+    /// Explicitly lock the matching row in the user_collections table. Read
+    /// locks do `SELECT ... LOCK IN SHARE MODE` and write locks do `SELECT
+    /// ... FOR UPDATE`.
+    ///
+    /// In theory it would be possible to use serializable transactions rather
+    /// than explicit locking, but our ops team have expressed concerns about
+    /// the efficiency of that approach at scale.
     async fn lock_for_read(
         &mut self,
         params: params::LockCollection,
-    ) -> Result<results::LockCollection, Self::Error> {
-        todo!()
+    ) -> DbResult<results::LockCollection> {
+        let user_id = params.user_id.legacy_id as i64;
+        let collection_id = self
+            .get_collection_id(&params.collection)
+            .await
+            .or_else(|e| {
+                if e.is_collection_not_found() {
+                    // If the collection doesn't exist, we still want to start a
+                    // transaction, so it will continue to not exist.
+                    Ok(0)
+                } else {
+                    Err(e)
+                }
+            })?;
+
+        // If we already have a read or write lock then it's safe to
+        // use it as-is.
+        if self
+            .session
+            .coll_locks
+            .contains_key(&(user_id as u32, collection_id))
+        {
+            return Ok(());
+        }
+
+        // Lock db.
+        self.begin(false).await?;
+
+        let modified = user_collections::table
+            .select(user_collections::modified)
+            .filter(user_collections::user_id.eq(user_id))
+            .filter(user_collections::collection_id.eq(collection_id))
+            .for_share()
+            .first(&mut self.conn)
+            .await
+            .optional()?;
+
+        if let Some(modified) = modified {
+            let modified = SyncTimestamp::from_i64(modified)?;
+            self.session
+                .coll_modified_cache
+                .insert((user_id as u32, collection_id), modified);
+        }
+        self.session
+            .coll_locks
+            .insert((user_id as u32, collection_id), CollectionLock::Read);
+        Ok(())
```

Comment:
This function is identical to the one in `syncstorage-mysql/src/db/db_impl.rs`.  Do we need two copies of this function?

(The MySQL copy also has [two](https://github.com/mozilla-services/syncstorage-rs/blob/70a7854b8fe1257320656efa933b21445e3ba7a3/syncstorage-mysql/src/db/db_impl.rs#L79) [questions](https://github.com/mozilla-services/syncstorage-rs/blob/70a7854b8fe1257320656efa933b21445e3ba7a3/syncstorage-mysql/src/db/db_impl.rs#L81) in the comments I'm very interested in.)

Reply:
The second question I've covered here: https://github.com/mozilla-services/syncstorage-rs/pull/1891#issuecomment-3556052449

The first question sounds like more of a TODO/bug, we shouldn't be casting a u64 to a u32 here (or in mysql)! Let's change the type in the `coll_` maps to match `legacy_id`'s. We could even put the entire `UserIdentifier` in there instead, as long as we're sure it eq/hashes correctly for such usage.